dockstore-webservice/src/main/java/io/dockstore/webservice/languages/CWLHandler.java

```diff
@@ -896,4 +900,293 @@ public VersionTypeValidation validateToolSet(Set<SourceFile> sourcefiles, String
+                if (supportsMixin(version)) {
+                    String mixinPath = findString(MIXIN_KEYS, map);
+                    if (mixinPath != null) {
+                        Object mixin = loadFileAndPreprocess(resolvePath(mixinPath, currentPath), emptyMap(), version, depth);
+                        if (mixin instanceof Map) {
+                            removeKey(MIXIN_KEYS, map);
+                            applyMixin(map, (Map<String, Object>)mixin);
+                        }
+                    }
+                }
+
+                // Process each value of the Map
+                preprocessMapValues(map, currentPath, version, depth);
+
+            } else if (cwl instanceof List) {
+
+                // Process each value of the List
+                preprocessListValues((List<Object>)cwl, currentPath, version, depth);
+
+            }
+
+            return cwl;
+        }
+
+        private void preprocessMapValues(Map<String, Object> cwl, String currentPath, String version, int depth) {
+
+            // Convert "run: {$import: <file>}" to "run: <file>"
+            Object runValue = cwl.get("run");
+            if (runValue instanceof Map) {
+                String importValue = findString(IMPORT_KEYS, (Map<String, Object>)runValue);
+                if (importValue != null) {
+                    cwl.put("run", importValue);
+                }
+            }
+
+            // Preprocess all values in the map
+            cwl.replaceAll((k, v) -> preprocess(v, currentPath, version, depth));
+
+            // Expand "run: <file>", leaving it unchanged if the file does not exist
+            runValue = cwl.get("run");
+            if (runValue instanceof String) {
+                String runPath = (String)runValue;
+                cwl.put("run", loadFileAndPreprocess(resolvePath(runPath, currentPath), runPath, version, depth));
+            }
+        }
+
+        private void preprocessListValues(List<Object> cwl, String currentPath, String version, int depth) {
+            cwl.replaceAll(v -> preprocess(v, currentPath, version, depth));
+        }
+
+        private boolean isEntry(Map<String, Object> cwl) {
+            String c = (String)cwl.get("class");
+            return Objects.equals(c, "CommandLineTool") || Objects.equals(c, "ExpressionTool") || Objects.equals(c, "Workflow");
+        }
+
+        private String setUniqueIdIfAbsent(Map<String, Object> entryCwl) {
+            String currentId = (String)entryCwl.get("id");
+            if (currentId == null || idToPath.containsKey(currentId)) {
+                entryCwl.put("id", java.util.UUID.randomUUID().toString());
+            }
+            return (String)entryCwl.get("id");
+        }
+
+        private boolean supportsMixin(String version) {
+            return version != null && version.startsWith("v1.0");
+        }
+
+        private String stripLeadingSlash(String value) {
```

Comment:
commons has this function which also handles the very weird case of multiple leading slashes (if that's even possible)